packages/react-components/react-table/src/hooks/useSelection.ts

```diff
@@ -1,14 +1,38 @@
-import { GetRowIdInternal, SelectionStateInternal, UseTableOptions } from './types';
-import { useMultipleSelection } from './useMultipleSelection';
-import { useSingleSelection } from './useSingleSelection';
+import * as React from 'react';
+import { useEventCallback, usePrevious } from '@fluentui/react-utilities';
+import { createSelectionManager } from './selectionManager';
+import { GetRowIdInternal, RowId, SelectionMode, SelectionStateInternal } from './types';
 
 export function useSelection<TItem>(
-  selectionMode: UseTableOptions<TItem>['selectionMode'],
+  selectionMode: SelectionMode,
   items: TItem[],
   getRowId: GetRowIdInternal<TItem>,
 ): SelectionStateInternal {
-  const multipleSelectionState = useMultipleSelection(items, getRowId);
-  const singleSelectionState = useSingleSelection();
+  const prevSelectionMode = usePrevious(selectionMode);
+  const [selected, setSelected] = React.useState(() => new Set<RowId>());
+  const [selectionManager, setSelectionManager] = React.useState(() =>
+    createSelectionManager(selectionMode, setSelected),
+  );
 
-  return selectionMode === 'multiselect' ? multipleSelectionState : singleSelectionState;
+  React.useEffect(() => {
+    if (prevSelectionMode !== selectionMode) {
+      setSelectionManager(createSelectionManager(selectionMode, setSelected));
+    }
+  }, [selectionMode, prevSelectionMode]);
```

Comment:
_not blocking:_ Let's pair on creating `useSingletone` or something like that as people who are not aware about Strict mode issues will be confused.

Reply:
yeah we even have a task for it, but not sure how much of a priority 18 is yet #24085